### PR TITLE
sgx: add new plugin.

### PR DESCRIPTION
**Pull request by ipuustin**
Add a new plugin for exposing SGX resources. Also, add a NFD source hook for counting SGX EPC memory.

sgx-resource-model

### CODE REVIEW

**Comment by ailin258** [comment]
in order to support inclavare-containers- https://github.com/alibaba/inclavare-containers
alibaba has developed a SGX device plugin for Kubernetes - https://github.com/AliyunContainerService/sgx-device-plugin
just for checking...

**Comment by mythi** [comment]
@ailin258 thanks! I've seen the implementation. This work takes a slightly different approach for EPC resources and also targets the upstream Linux driver only. We're also looking to add more than just the plugin (see #379 and #380)

**Comment by ailin258** [comment]
@mythi  good to know you have seen this already

**Comment by mythi** [comment]
@ipuustin one idea. as we are going to have a webhook anyways, could the "user interface" be `sgx.intel.com/epc` only and the `sgx.intel.com/enclave` gets added by the webhook. similarly, the provision part could be added there based on some webhook validation criteria

**Comment by mythi** [comment]
@rojkov @bart0sh PTAL? we'd like to merge this (WIP label added to the readme)

File: cmd/sgx_plugin/sgx_plugin.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+func (dp *devicePlugin) Scan(notifier dpapi.Notifier) error {
+	devTree, err := dp.scan()
+	if err != nil {
+		return err
+	}
+	notifier.Notify(devTree)
+
+	// Wait forever to prevent manager run loop from exiting.
+	<-dp.scanDone
```
Comment (rojkov) [inline cmd/sgx_plugin/sgx_plugin.go:64]: Is it going to be one-time scan always? I'm thinking of the case for loading/unloading the kernel driver.

File: cmd/sgx_plugin/sgx_plugin.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+func getDefaultPodCount(nCPUs uint) uint {
+	// By default we provide as many enclave resources as there can be pods
+	// running on the node. The problem is that this value is configurable
+	// either via "--pods-per-core" or "--max-pods" kubelet options. We get the
+	// limit by multiplying the number of cores in the system with env variable
+	// "PODS_PER_CORE".
+
+	podsPerCore := uint(defaultPodsPerCore)
```
Comment (rojkov) [inline cmd/sgx_plugin/sgx_plugin.go:103]: Probably it would be safer to use a typed global constant for `defaultPodsPerCore`.

**Review by rojkov — COMMENTED** [review]
Looks good! Just one question to clarify.

File: cmd/sgx_plugin/sgx_plugin.go (hunk 1 above)
Comment (ipuustin) [reply cmd/sgx_plugin/sgx_plugin.go:64]: I believe that SGX driver can't be built as module, so it shouldn't be possible to load or unload.

File: cmd/sgx_plugin/sgx_plugin.go (hunk 2 above)
Comment (ipuustin) [reply cmd/sgx_plugin/sgx_plugin.go:103]: Ok, fixed.

File: cmd/sgx_plugin/sgx_plugin.go (hunk 1 above)
Comment (mythi) [reply cmd/sgx_plugin/sgx_plugin.go:64]: correct, it can't be built as a module (`CONFIG_INTEL_SGX` is `bool`)

**Review by rojkov — APPROVED** [review]

**Review by mythi — APPROVED** [review]

**Pull request merged**